mp/src/game/shared/momentum/run/mom_run_safeguards.cpp

```diff
@@ -0,0 +1,180 @@
+
+#ifdef CLIENT_DLL
+#include "input.h"
+#endif
+
+#include "tier0/memdbgon.h"
+
+#define MOM_DOUBLEPRESS_MAXTIME 0.5f
+
+static MAKE_CONVAR(mom_run_safeguard_practicemode, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for enabling practice mode during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_FIRST, RUN_SAFEGUARD_MODE_LAST);
+
+static MAKE_CONVAR(mom_run_safeguard_restart, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for restarting the map during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_FIRST, RUN_SAFEGUARD_MODE_LAST);
+
+static MAKE_CONVAR(mom_run_safeguard_saveloc_tele, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for teleporting to a saved location during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_FIRST, RUN_SAFEGUARD_MODE_LAST);
+
+static MAKE_CONVAR(mom_run_safeguard_chat_open, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for opening chat during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_FIRST, RUN_SAFEGUARD_MODE_LAST);
+
+MomRunSafeguards::MomRunSafeguards()
+    : m_pLocalPlayer(nullptr)
+{
+    m_arrpszAction[RUN_SAFEGUARD_PRACTICEMODE] = "enable practice mode";
+    m_arrpszAction[RUN_SAFEGUARD_RESTART] = "restart map";
+    m_arrpszAction[RUN_SAFEGUARD_SAVELOC_TELE] = "teleport to saveloc";
+    m_arrpszAction[RUN_SAFEGUARD_CHAT_OPEN] = "open chat";
+
+    for (int i = 0; i < RUN_SAFEGUARD_COUNT; i++)
+    {
+        m_szLastTimesPressed[i] = 0.0f;
+        m_szDoublePressSafeguards[i] = true;
+    }
+}
+
+bool MomRunSafeguards::IsSafeguarded(RunSafeguardType_t type)
+{
+    return IsSafeguarded(type, GetModeFromType(type));
+}
+
+bool MomRunSafeguards::IsSafeguarded(RunSafeguardType_t type, RunSafeguardMode_t mode)
+{
+    CMomentumPlayer *pPlayer;
+#ifdef CLIENT_DLL
+    pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+#elif GAME_DLL
+    pPlayer = CMomentumPlayer::GetLocalPlayer();
+#endif
+    if (!pPlayer)
+        return false; 
+
+    CMomRunEntityData *pEntData;
+#ifdef CLIENT_DLL
+    pEntData = pPlayer->GetCurrentUIEntData();
+#elif GAME_DLL
+    pEntData = pPlayer->GetRunEntData();
+#endif
+    if (!pEntData)
+        return false;
+
+    if (!pEntData->m_bTimerRunning || pPlayer->IsObserver())
+        return false; 
+
+    m_pLocalPlayer = pPlayer;
+
+    switch (mode)
+    {
+    case RUN_SAFEGUARD_MODE_MOVEMENTKEYS:
+        return IsMovementKeysSafeguarded(type);
+    case RUN_SAFEGUARD_MODE_DOUBLEPRESS:
+        return IsDoublePressSafeguarded(type);
+    default:
+        return false;
+    }
```

Comment:
Due to the mode being stored in the CRunSafeguard class, this would just move to calling like `pSafeguard->IsSafeguarded` and having the safeguard class itself do this switch statement depending on its mode.